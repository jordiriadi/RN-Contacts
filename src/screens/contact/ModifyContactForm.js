import React from 'react';
import PropTypes from 'prop-types';
import {StyleSheet, View, Text, TextInput} from 'react-native';
import { Avatar } from 'react-native-elements';
import Button from '../../components/Button';
import Styles from '../../utils/Styles';

class ModifyContactForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      fieldValues: {},
    };
  }

  contactFields = [
    {
      fieldName: 'photo',
      type: 'image',
      text: 'Photo',
    },
    {
      fieldName: 'firstName',
      type: 'default',
      text: 'First Name',
    },
    {
      fieldName: 'lastName',
      type: 'default',
      text: 'Last Name',
    },
    {
      fieldName: 'age',
      type: 'numeric',
      text: 'Age',
    },
  ];

  componentDidMount() {
    if (this.props.modify === "update") {
      this.setState({
        fieldValues: {...this.props.value}
      });
    }
  }

  handleChange = (fieldName, type) => value => {
    if (type === 'numeric') {
      const _value = value.replace(/[^0-9]/g, '');
      return this.handleSetValue(fieldName, _value);
    }

    return this, this.handleSetValue(fieldName, value);
  };

  handleSetValue = (fieldName, value) =>
    this.setState({
      fieldValues: Object.assign({}, this.state.fieldValues, {
        [fieldName]: value,
      }),
    });

  render() {
    const {onSave, isLoading} = this.props;
    const {fieldValues} = this.state;

    return (
      <View id="form-container" style={styles.container}>
        {this.contactFields.map(field => (
          <View id={`field-${field.fieldName}`} key={field.fieldName} style={styles.fieldContainer}>
            {field.type === "image" ? (
              <View style={styles.avatarContainer}>
              <Avatar
                rounded
                size="large"
                icon={{name: 'camera', color: 'black', type: 'font-awesome'}}
                overlayContainerStyle={{backgroundColor: 'orange'}}
                source={{
                  uri:fieldValues.photo || "N/A"
                }}
                showAccessory={true}
                title={fieldValues.initial}
              />
              </View>
            ) : (
              <React.Fragment>
                <Text>{field.text}</Text>
                <TextInput
                  id={`input-${field.fieldName}`}
                  keyboardType={field.type}
                  style={styles.textInput}
                  onChangeText={this.handleChange(field.fieldName, field.type)}
                  value={fieldValues[field.fieldName]}
                />
              </React.Fragment>
            )}
          </View>
        ))}
        <View style={styles.buttonContainer}>
          <View style={styles.button}>
            <Button type="solid" disabled={isLoading}>
              Cancel
            </Button>
          </View>
          <View  style={styles.button}>
            <Button id="btn-save" disabled={isLoading} onPress={() => onSave(this.state.fieldValues)}>
              Save
            </Button>
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "white",
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    padding: Styles.DefaultPadding,
    marginTop: Styles.DefaultMarginTop,
  },
  textInput: {
    height: 30,
    padding: 3,
    borderColor: 'gray',
    borderWidth: 0.8,
  },
  avatarContainer: {
    flex: 1,
    alignItems: 'center',
    marginTop: Styles.DefaultMargin*-6,
    marginBottom: Styles.DefaultMargin*8
  },
  fieldContainer: {
    paddingTop: 10,
  },
  buttonContainer: {
    // flex: 1,
    flex: 1,
    flexDirection: 'row',
    position: 'absolute',
    left: 0, right: 0, bottom: 0
  },
  button: {
    flex: 1,
  }
});

ModifyContactForm.propTypes = {
  modify: PropTypes.oneOf(["update", "create"]),
  value: PropTypes.object,
  onSave: PropTypes.func,
  isLoading: PropTypes.bool
}

export default ModifyContactForm;
