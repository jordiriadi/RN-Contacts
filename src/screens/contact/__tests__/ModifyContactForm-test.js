import 'react-native';
import React from 'react';
import {shallow} from 'enzyme';

import ModifyContactForm from '../ModifyContactForm';

const mockData = {
  "age": "33",
  "firstName": "LukeTheSecond",
  "id": "b3abd640-b22g-11e8-b02f-cbfa15db428b",
  "lastName": "Skywalker", 
  "initial": "SL",
  "photo": "N/A"
}

describe('Contact  Detail Form', () => {  
  it('renders correctly', () => {
  
    const wrapper = shallow(<ModifyContactForm contact={mockData}/>);

    expect(wrapper).toMatchSnapshot();
  });

  
  it('can add contact correctly', () => {
    const wrapper = shallow(<ModifyContactForm modify="create" contact={mockData} />);

    wrapper.setState({fieldValues: mockData});
    expect(wrapper.state().fieldValues).toEqual(mockData);

    wrapper.find('#input-age').simulate('changeText', '23');
    wrapper.find('#input-firstName').simulate('changeText', 'first name');
    wrapper.find('#btn-save').simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
})