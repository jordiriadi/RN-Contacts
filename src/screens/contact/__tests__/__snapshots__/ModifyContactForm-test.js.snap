// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact  Detail Form renders correctly 1`] = `ShallowWrapper {}`;
